core/hashstorage.go

```diff
@@ -0,0 +1,153 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Package core implements the Ethereum consensus protocol.
+
+package core
+
+import (
+	"bytes"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// For EVM execution, we need around 256 items. We add a few more to allow reorgs.
+// For mainnet, a couple more would suffice, but a few more added for
+// testnets/private nets
+const HashStorageElems = 288
+const HashStorageDataSize = 32 * HashStorageElems
+
+// HashStorage implements a storage for chains of hashes, intended to be used for quick lookup of blockhashes.
+// Internally, it uses an array of 8192 bytes to store 256 hashes in a circular buffer.
+// It enforces that all hashes added have a contiguous parent-child relation, and supports rollbacks
+// It is not thread-safe.
+type HashStorage struct {
+
+	// The data holds 256 32-byte hashes. The hashes are sequential, but also a
+	// circular buffer.
+	// The `head` points to the position of the latest hash.
+	// The parent, if present, is located 32 bytes back.
+	// [.., .., ..., head-2 , head-1, head, oldest, ... ]
+	data [HashStorageDataSize]byte
+
+	head uint64 // index of hash for the head block
+
+	headNumber uint64 // The blocknumber for head (the most recent block)
+	tailNumber uint64 // The blocknumber for tail (the oldest block)
+}
+
+// Create a new storage with a header in it.
+func NewHashStorage(hdr *types.Header) *HashStorage {
+
+	// Since we take a header here, a hash storage can never be empty.
+	// This makes things easier later on (in Set)
+	hs := &HashStorage{}
+	hs.head = 0
+	hs.headNumber = hdr.Number.Uint64()
+	hs.tailNumber = hs.headNumber
+	hash := hdr.Hash()
+	copy(hs.data[0:], hash[:])
+	return hs
+
+}
+
+// Get locates the hash for the requested number, and copies the data into the
+// provided big.Int -- zeroing it in case the hash is not found
+func (hs *HashStorage) Get(number uint64) (common.Hash, bool) {
+	var h common.Hash
+	if !hs.has(number) {
+		return h, false
+	}
+	distance := hs.headNumber - number
+	idx := (hs.head + HashStorageDataSize - distance*32) % HashStorageDataSize
+	copy(h[:], hs.data[idx:idx+32])
+	return h, true
+}
+
+// has returns if the storage has a hash for the given number
+func (hs *HashStorage) has(number uint64) bool {
+	return number <= hs.headNumber && number >= hs.tailNumber
+}
+
+// Is checks if the hash at the given number matches the expected
```

Comment:
It was not immediately apparent to me what this function did by it's name. Consider using `Has` or `Contains`.

Reply:
I agree